### ISSUE TITLE
Error "CONTAINER_JAR_PATTERN"  with Gretty 4.0.0

### ISSUE THREAD

**Issue by hkujath22** [gretty-gradle-plugin/gretty#248]
Hi,

when i try to use the plugin with "jettyrun" i got following error. Am i doing something wrong or what is the problem?

```
> Task :prepareInplaceWebAppFolder UP-TO-DATE
> Task :createInplaceWebAppFolder
> Task :compileJava UP-TO-DATE
> Task :processResources UP-TO-DATE
> Task :classes UP-TO-DATE
> Task :prepareInplaceWebAppClasses UP-TO-DATE
> Task :prepareInplaceWebApp
15:35:04.345 [main] DEBUG o.akhikhl.gretty.JettyServerManager - Jetty 11.0.3 starting.
15:35:12.574 [main] ERROR o.akhikhl.gretty.JettyServerManager - Error starting server
groovy.lang.MissingPropertyException: No such property: CONTAINER_JAR_PATTERN for class: org.eclipse.jetty.webapp.MetaInfConfiguration
	at groovy.lang.MetaClassImpl.invokeStaticMissingProperty(MetaClassImpl.java:1019)
	at groovy.lang.MetaClassImpl.getProperty(MetaClassImpl.java:1963)
	at groovy.lang.MetaClassImpl.getProperty(MetaClassImpl.java:1938)
	at groovy.lang.MetaClassImpl.getProperty(MetaClassImpl.java:3843)
	at org.codehaus.groovy.runtime.callsite.ClassMetaClassGetPropertySite.getProperty(ClassMetaClassGetPropertySite.java:50)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callGetProperty(AbstractCallSite.java:329)
	at org.akhikhl.gretty.JettyConfigurerImpl.createWebAppContext(JettyConfigurerImpl.groovy:206)
	at org.akhikhl.gretty.JettyConfigurer$createWebAppContext$5.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:148)
	at org.akhikhl.gretty.JettyServerConfigurer.createContext(JettyServerConfigurer.groovy:74)
	at org.akhikhl.gretty.JettyServerConfigurer.createContext(JettyServerConfigurer.groovy)
	at org.akhikhl.gretty.JettyServerConfigurer$createContext$0.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:157)
	at org.akhikhl.gretty.JettyServerManager.startServer(JettyServerManager.groovy:58)
	at org.akhikhl.gretty.ServerManager$startServer$0.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:139)
	at org.akhikhl.gretty.Runner.run(Runner.groovy:141)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.codehaus.groovy.runtime.callsite.PlainObjectMetaMethodSite.doInvoke(PlainObjectMetaMethodSite.java:43)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoCachedMethodSiteNoUnwrapNoCoerce.invoke(PogoMetaMethodSite.java:193)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.call(PogoMetaMethodSite.java:73)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:130)
	at org.akhikhl.gretty.Runner.main(Runner.groovy:54)
15:35:12.584 [main] INFO  o.a.gretty.JettyServerStartInfo - Jetty 11.0.3 started and listening on port 8080
15:35:12.597 [main] DEBUG org.akhikhl.gretty.ServiceProtocol - ServiceProtocol.send(63038, {"status":"error starting server","host":"localhost","httpPort":8080,"contexts":[],"error":true,"errorMessage":"No such property: CONTAINER_JAR_PATTERN for class: org.eclipse.jetty.webapp.MetaInfConfiguration","stackTrace":"groovy.lang.MissingPropertyException: No such property: CONTAINER_JAR_PATTERN for class: org.eclipse.jetty.webapp.MetaInfConfiguration\r\n\tat groovy.lang.MetaClassImpl.invokeStaticMissingProperty(MetaClassImpl.java:1019)\r\n\tat groovy.lang.MetaClassImpl.getProperty(MetaClassImpl.java:1963)\r\n\tat groovy.lang.MetaClassImpl.getProperty(MetaClassImpl.java:1938)\r\n\tat groovy.lang.MetaClassImpl.getProperty(MetaClassImpl.java:3843)\r\n\tat org.codehaus.groovy.runtime.callsite.ClassMetaClassGetPropertySite.getProperty(ClassMetaClassGetPropertySite.java:50)\r\n\tat org.codehaus.groovy.runtime.callsite.AbstractCallSite.callGetProperty(AbstractCallSite.java:329)\r\n\tat org.akhikhl.gretty.JettyConfigurerImpl.createWebAppContext(JettyConfigurerImpl.groovy:206)\r\n\tat org.akhikhl.gretty.JettyConfigurer$createWebAppContext$5.call(Unknown Source)\r\n\tat org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)\r\n\tat org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)\r\n\tat org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:148)\r\n\tat org.akhikhl.gretty.JettyServerConfigurer.createContext(JettyServerConfigurer.groovy:74)\r\n\tat org.akhikhl.gretty.JettyServerConfigurer.createContext(JettyServerConfigurer.groovy)\r\n\tat org.akhikhl.gretty.JettyServerConfigurer$createContext$0.call(Unknown Source)\r\n\tat org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)\r\n\tat org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)\r\n\tat org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:157)\r\n\tat org.akhikhl.gretty.JettyServerManager.startServer(JettyServerManager.groovy:58)\r\n\tat org.akhikhl.gretty.ServerManager$startServer$0.call(Unknown Source)\r\n\tat org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)\r\n\tat org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)\r\n\tat org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:139)\r\n\tat org.akhikhl.gretty.Runner.run(Runner.groovy:141)\r\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\r\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\r\n\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\r\n\tat java.base/java.lang.reflect.Method.invoke(Method.java:566)\r\n\tat org.codehaus.groovy.runtime.callsite.PlainObjectMetaMethodSite.doInvoke(PlainObjectMetaMethodSite.java:43)\r\n\tat org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoCachedMethodSiteNoUnwrapNoCoerce.invoke(PogoMetaMethodSite.java:193)\r\n\tat org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.call(PogoMetaMethodSite.java:73)\r\n\tat org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)\r\n\tat org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)\r\n\tat org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:130)\r\n\tat org.akhikhl.gretty.Runner.main(Runner.groovy:54)\r\n"})
> Task :jettyRun FAILED
5 actionable tasks: 2 executed, 3 up-to-date

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':jettyRun'.
> java.lang.Exception: No such property: CONTAINER_JAR_PATTERN for class: org.eclipse.jetty.webapp.MetaInfConfiguration
```

**Comment by f4lco** [comment]
🤔Interesting. Personally I'll be able to allocate time towards the end of next week to investigate further.

Would you mind to post snippets of your Gretty configuration? Which features are you using? We test many configuration options extensively, I wonder why this fault does not appear when running the integration tests.

**Comment by hkujath** [comment]
Hi. No problem. Its not urgent. I have no specific gretty configuration.

Here is my build.gradle. 

`plugins {
    id 'org.springframework.boot' version '2.6.3'
    id 'io.spring.dependency-management' version '1.0.11.RELEASE'
    id 'java'
    id 'war'
}

apply plugin: 'war'
apply from: 'https://raw.github.com/gretty-gradle-plugin/gretty/master/pluginScripts/gretty.plugin'



group = 'de.mycompany'
version = '0.0.1-SNAPSHOT'
sourceCompatibility = '11'
targetCompatibility = '11'




configurations {
    compileOnly {
        extendsFrom annotationProcessor
    }
}

repositories {
    mavenCentral()
}

dependencies {
    implementation 'org.springframework.boot:spring-boot-starter'
    implementation 'org.springframework:spring-webmvc'
    compileOnly('jakarta.validation:jakarta.validation-api:3.0.1')
    compileOnly('jakarta.servlet:jakarta.servlet-api:5.0.0')
    implementation('org.eclipse.persistence:org.eclipse.persistence.jpa:3.0.2')


    compileOnly 'org.projectlombok:lombok'
    annotationProcessor 'org.projectlombok:lombok'

    compileOnly 'org.projectlombok:lombok:1.18.22'
    annotationProcessor 'org.projectlombok:lombok:1.18.22'

    testImplementation 'org.springframework.boot:spring-boot-starter-test'

}

tasks.named('test') {
    useJUnitPlatform()
}
`

**Comment by f4lco** [comment]
Hi @hkujath @hkujath22 thank you for the `build.gradle`, only with those instructions I was able to replicate the issue. 

I conducted a Gradle build scan and found the following dependency configuration:

<img src=https://user-images.githubusercontent.com/29141485/152774300-ad26f75c-2c72-4e01-bf00-002937d34cfb.png width=300/>

So the Spring Dependency Management Plugin forces the 'wrong' dependency versions for use with Gretty 4.
- Gretty 4 supports Jetty 11 onwards, and does not support Jetty 9.x. You'll need to lift that dependency constraint if you really want to use Gretty 4 / Jetty 11.
- Gretty 3 does support Jetty 9.x, maybe you'll want to give Gretty 3.0.7 a try.
- When just plugging in Gretty 3.x, Gretty will complain about the missing `GafferConfigurator`. If you stick with that version of the Spring Dependency Management Plugin, you'll need to configure the dependency management plugin to downgrade Logback to 1.2.8 or below, because release 1.2.9 removed the ability to use Groovy to configure logging, a feature of Logback that Gretty 3.x currently relies on.
 
I hope this addresses the issues you are experiencing with Gretty.
For being able to use the latest Logback version with Gretty I opened #249. It's up for grabs, contributions welcome 😉
